### ISSUE TITLE
IdleTimeout Error

### ISSUE THREAD

**Issue by felahi** [brettwooldridge/HikariCP#610]
I am overriding idleTimeout value to <property name="hibernate.hikari.idleTimeout" value="60000"/>. I am testing using JMeter. I executed 50 queries. Max connections created = 10 which is fine. But after execution none of the connections timeout as per idleTimout value they took maxLifeTime to timeout.

**Comment by brettwooldridge** [comment]
@felahi Can you enable debug logging, and post the end of the log (last 20-30 lines) after the execution run?

**Comment by felahi** [comment]
ok doing

**Comment by felahi** [comment]
17:51:53,041 DEBUG HikariConnectionProvider:configure:84 - Configuring HikariCP
17:51:53,049 DEBUG HikariConfig:logConfiguration:838 - HikariPool-1 - configuration:
17:51:53,051 DEBUG HikariConfig:logConfiguration:854 - allowPoolSuspension.............false
17:51:53,051 DEBUG HikariConfig:logConfiguration:854 - autoCommit......................true
17:51:53,051 DEBUG HikariConfig:logConfiguration:854 - catalog.........................null
17:51:53,053 DEBUG HikariConfig:logConfiguration:854 - connectionInitSql...............null
17:51:53,054 DEBUG HikariConfig:logConfiguration:854 - connectionTestQuery.............null
17:51:53,055 DEBUG HikariConfig:logConfiguration:854 - connectionTimeout...............180000
17:51:53,055 DEBUG HikariConfig:logConfiguration:854 - dataSource......................null
17:51:53,056 DEBUG HikariConfig:logConfiguration:854 - dataSourceClassName............."com.mysql.jdbc.jdbc2.optional.MysqlDataSource"
17:51:53,056 DEBUG HikariConfig:logConfiguration:854 - dataSourceJNDI..................null
17:51:53,057 DEBUG HikariConfig:logConfiguration:854 - dataSourceProperties............{user=umairullah, password=<masked>, url=jdbc:mysql://qa-3.cmsz241itecr.eu-west-1.rds.amazonaws.com:3306/careem}
17:51:53,057 DEBUG HikariConfig:logConfiguration:854 - driverClassName.................null
17:51:53,057 DEBUG HikariConfig:logConfiguration:854 - healthCheckProperties...........{}
17:51:53,058 DEBUG HikariConfig:logConfiguration:854 - healthCheckRegistry.............null
17:51:53,058 DEBUG HikariConfig:logConfiguration:854 - idleTimeout.....................60000
17:51:53,059 DEBUG HikariConfig:logConfiguration:854 - initializationFailFast..........true
17:51:53,059 DEBUG HikariConfig:logConfiguration:854 - isolateInternalQueries..........false
17:51:53,059 DEBUG HikariConfig:logConfiguration:854 - jdbc4ConnectionTest.............false
17:51:53,059 DEBUG HikariConfig:logConfiguration:854 - jdbcUrl.........................null
17:51:53,060 DEBUG HikariConfig:logConfiguration:854 - leakDetectionThreshold..........0
17:51:53,060 DEBUG HikariConfig:logConfiguration:854 - maxLifetime.....................1800000
17:51:53,060 DEBUG HikariConfig:logConfiguration:854 - maximumPoolSize.................10
17:51:53,061 DEBUG HikariConfig:logConfiguration:854 - metricRegistry..................null
17:51:53,061 DEBUG HikariConfig:logConfiguration:854 - metricsTrackerFactory...........null
17:51:53,061 DEBUG HikariConfig:logConfiguration:854 - minimumIdle.....................5
17:51:53,061 DEBUG HikariConfig:logConfiguration:854 - password........................<masked>
17:51:53,062 DEBUG HikariConfig:logConfiguration:854 - poolName........................"HikariPool-1"
17:51:53,062 DEBUG HikariConfig:logConfiguration:854 - readOnly........................false
17:51:53,062 DEBUG HikariConfig:logConfiguration:854 - registerMbeans..................false
17:51:53,064 DEBUG HikariConfig:logConfiguration:854 - scheduledExecutorService........null
17:51:53,064 DEBUG HikariConfig:logConfiguration:854 - threadFactory...................null
17:51:53,064 DEBUG HikariConfig:logConfiguration:854 - transactionIsolation............null
17:51:53,065 DEBUG HikariConfig:logConfiguration:854 - username........................null
17:51:53,065 DEBUG HikariConfig:logConfiguration:854 - validationTimeout...............5000

17:51:55,749 DEBUG HikariConnectionProvider:configure:94 - HikariCP Configured
17:51:57,540 DEBUG HikariPool:createPoolEntry:454 - HikariPool-1 - Added connection com.mysql.jdbc.JDBC4Connection@596f801

**Comment by felahi** [comment]
In the following way I have enable logger
<logger name="com.zaxxer.hikari">
        <level value="dubug"/>
    </logger>

**Comment by felahi** [comment]
<dependency>
            <groupId>com.zaxxer</groupId>
            <artifactId>HikariCP</artifactId>
            <version>2.4.5</version>
            <scope>compile</scope>
        </dependency>

**Comment by brettwooldridge** [comment]
Great, now can you append the log after the JMeter run finishes?  You should see log messages like:
```
2016-03-28 12:29:31,895 [Pool housekeeper] DEBUG com.zaxxer.hikari.pool.HikariPool - Read-Write-Pool - Pool stats (total=10, active=1, idle=9, waiting=0)
2016-03-28 12:30:01,895 [housekeeper] DEBUG com.zaxxer.hikari.pool.HikariPool - Read-Write-Pool - Before cleanup stats (total=11, active=0, idle=11, waiting=0)
2016-03-28 12:30:01,896 [housekeeper] DEBUG com.zaxxer.hikari.pool.HikariPool - Read-Write-Pool - After cleanup stats (total=10, active=0, idle=10, waiting=0)
```

**Comment by felahi** [comment]
I am not getting these logs :(

**Comment by felahi** [comment]
18:12:15,418 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=88)
18:12:15,419 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:15,469 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=88)
18:12:15,469 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:15,543 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=90)
18:12:15,543 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:15,565 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=89)
18:12:16,002 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=96)
18:12:16,003 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:16,674 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=92)
18:12:16,674 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:18,139 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=90)
18:12:18,139 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:20,546 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=98)
18:12:20,547 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:20,757 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=96)
18:12:20,757 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:24,013 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=99)
18:12:24,013 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:24,350 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=98)
18:12:24,351 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:28,590 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=99)
18:12:28,592 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.

**Comment by felahi** [comment]
18:12:15,418 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=88)
18:12:15,419 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:15,469 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=88)
18:12:15,469 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:15,543 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=90)
18:12:15,543 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:15,565 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=89)
18:12:16,002 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=96)
18:12:16,003 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:16,674 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=92)
18:12:16,674 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:18,139 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=90)
18:12:18,139 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180002ms.
18:12:20,546 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=98)
18:12:20,547 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:20,757 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=96)
18:12:20,757 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:24,013 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=99)
18:12:24,013 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180000ms.
18:12:24,350 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=98)
18:12:24,351 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.
18:12:28,590 DEBUG HikariPool:logPoolState:387 - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=99)
18:12:28,592 ERROR SqlExceptionHelper:logExceptions:146 - HikariPool-1 - Connection is not available, request timed out after 180001ms.

**Comment by brettwooldridge** [comment]
@felahi That looks like a possible connection leak to me.  Can you set ``leakDetectionThreshold=30000`` and see if a leak is reported?

**Comment by felahi** [comment]
@brettwooldridge Thats logs only appera if a run a very slow and complex query that took  more time that connectionTimeout.

**Comment by felahi** [comment]
I am bit confused that inspite of giving custom idleTimeout value i.e <property name="hibernate.hikari.idleTimeout" value="60000"/> connection is not reseeting like C3p0 or back to minPoolsize

**Comment by felahi** [comment]
- [ ] 18:32:03,335  WARN ProxyLeakTask:run:87 - Connection leak detection triggered for com.mysql.jdbc.JDBC4Connection@190823b8, stack trace follows
- [ ] java.lang.Exception: Apparent connection leak detected
- [ ] 	at com.zaxxer.hikari.hibernate.HikariConnectionProvider.getConnection(HikariConnectionProvider.java:106)
- [ ] 	at org.hibernate.internal.AbstractSessionImpl$NonContextualJdbcConnectionAccess.obtainConnection(AbstractSessionImpl.java:387)
- [ ] 	at org.hibernate.engine.jdbc.internal.LogicalConnectionImpl.obtainConnection(LogicalConnectionImpl.java:228)
- [ ] 	at org.hibernate.engine.jdbc.internal.LogicalConnectionImpl.getConnection(LogicalConnectionImpl.java:171)
- [ ] 	at org.hibernate.engine.transaction.internal.jdbc.JdbcTransaction.doBegin(JdbcTransaction.java:67)
- [ ] 	at org.hibernate.engine.transaction.spi.AbstractTransactionImpl.begin(AbstractTransactionImpl.java:162)
- [ ] 	at org.hibernate.internal.SessionImpl.beginTransaction(SessionImpl.java:1435)
- [ ] 	at org.hibernate.jpa.internal.TransactionImpl.begin(TransactionImpl.java:61)
- [ ] 	at com.careem.backend.persistence.TripRepository.getTripCount(TripRepository.java:138)
- [ ] 	at com.careem.backend.action.CBUserProfileAction.execute(CBUserProfileAction.java:39)
- [ ] 	at com.careem.controller.http.CBUserController.getUserProfile(CBUserController.java:29)
- [ ] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
- [ ] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
- [ ] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
- [ ] 	at java.lang.reflect.Method.invoke(Method.java:498)
- [ ] 	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:222)
- [ ] 	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
- [ ] 	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:110)
- [ ] 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:814)
- [ ] 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:737)
- [ ] 	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
- [ ] 	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
- [ ] 	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
- [ ] 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
- [ ] 	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
- [ ] 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
- [ ] 	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
- [ ] 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
- [ ] 	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
- [ ] 	at org.togglz.servlet.TogglzFilter.doFilter(TogglzFilter.java:100)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
- [ ] 	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:121)
- [ ] 	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
- [ ] 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
- [ ] 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
- [ ] 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
- [ ] 	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
- [ ] 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
- [ ] 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
- [ ] 	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
- [ ] 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
- [ ] 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:522)
- [ ] 	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1095)
- [ ] 	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:672)
- [ ] 	at org.apache.tomcat.util.net.AprEndpoint$SocketProcessor.doRun(AprEndpoint.java:2500)
- [ ] 	at org.apache.tomcat.util.net.AprEndpoint$SocketProcessor.run(AprEndpoint.java:2489)
- [ ] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
- [ ] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
- [ ] 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
- [ ] 	at java.lang.Thread.run(Thread.java:745)
- [ ] 18:32:04,551  INFO SSMCache:get:94 - Cache hit.

**Comment by brettwooldridge** [comment]
@felahi If you are using Hibernate >= 4.3.6 then you should be using their HikariCP connection provider, ``org.hibernate.hikaricp.internal.HikariCPConnectionProvider``.

Also, if you set your ``leakDetectionThreshold`` to a very long value -- for example 1 minute -- and you still get leak messages, then there is probably a configuration issue causing Spring or Hibernate to not return connections.

**Comment by felahi** [comment]
@brettwooldridge  

<version>4.3.2.Final</version>

**Comment by felahi** [comment]
@brettwooldridge  Now what shpuld you suggest ?

**Comment by felahi** [comment]
@brettwooldridge Can you please look at the configuration please, its urgent


                       <property name="hibernate.hikari.dataSource.url" value="${DATABASE_URL}"/>
            <property name="hibernate.hikari.dataSource.user" value="${DATABASE_USERNAME}"/>
            <property name="hibernate.hikari.dataSource.password" value="${DATABASE_PASSWORD}"/>
            <property name="hibernate.hikari.dataSourceClassName" value="com.mysql.jdbc.jdbc2.optional.MysqlDataSource"/>
            <property name="hibernate.connection.provider_class" value="com.zaxxer.hikari.hibernate.HikariConnectionProvider"/>
            <property name="hibernate.hikari.maxLifetime" value="1800000"/>
            <property name="hibernate.hikari.idleTimeout" value="60000"/>
            <property name="hibernate.hikari.connectionTimeout" value="180000" />
            <property name="hibernate.hikari.maximumPoolSize" value="10"/>
            <property name="hibernate.hikari.minimumIdle" value="5"/>
            <property name="hibernate.hikari.leakDetectionThreshold" value="30000" />

**Comment by brettwooldridge** [comment]
@felahi Can you try HikariCP 2.4.3 and let me know if the pool behavior is different?

**Comment by felahi** [comment]
@brettwooldridge same behavior.

**Comment by felahi** [comment]
@brettwooldridge What if there are so many queries executed and some queries hold connection and connection max pool size is limited then is there any property that handle this situation ?

**Comment by felahi** [comment]
@brettwooldridge is there any property like autoIncrement in C3P0 ?

**Comment by brettwooldridge** [comment]
Re: your previous question, whatever you configure as maximumPoolSize is the max, so if you run out of connections and your queries are running for a long time, then you have two choices.  1) increase the pool size, or 2) profile and optimize the slow queries.  I suggest that the latter is going to be more effective.  Adding more connections can typically slow things down by increasing the load on the DB.

Re: autoIncrement, HikariCP and C3P0 maintain idle connection levels in very different ways, so that property would not be particularly meaningful to HikariCP.

**Comment by felahi** [comment]
@brettwooldridge Example: I executed the command "SHOW FULL PROCESSLIST"  in MySql. Initially teher are is only one connection for my user from which I logged in. After that I created bulk of request from JMeter and connection reaches to max size i.e 10. After all requests are processed using those connections, none of the connection retire as per idleTimeout. They all took maxLifeTime to expire. Is it ok with HIkariCP or its an issue ?

**Comment by nitincchauhan** [comment]
@felahi
just in case if you are mis-informed, you need to "close" the connection (after done with processing)

**Comment by felahi** [comment]
@nitincchauhan  I have test it out idleTimout only reduces the connections from max to min like if we have 30 connections at max  and minIdle = 5 then it only reduces connection from 30 to 5 after idleTimout

**Comment by nitincchauhan** [comment]
behaviour is as documented. please read minimumIdle and other properties for better understanding here https://github.com/brettwooldridge/HikariCP#configuration-knobs-baby

**Comment by felahi** [comment]
@brettwooldridge If this issue has been fixed then when did this deployed ?

**Comment by brettwooldridge** [comment]
It is still unclear whether this change will address what you are experiencing.  This fixes an issue that was introduced between v2.4.4 and v2.4.5.  You said you tried version 2.4.3 and still observed idle connections not being closed, so I do have a doubt whether this will change anything.

Either way, we will be publishing this fix later today.

**Comment by felahi** [comment]
@brettwooldridge Working fine in 2.4.3
